shell/platform/windows/compositor_opengl_unittests.cc

```diff
@@ -0,0 +1,192 @@
+#include "flutter/shell/platform/windows/testing/engine_modifier.h"
+#include "flutter/shell/platform/windows/testing/flutter_windows_engine_builder.h"
+#include "flutter/shell/platform/windows/testing/mock_window_binding_handler.h"
+#include "flutter/shell/platform/windows/testing/windows_test.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace flutter {
+namespace testing {
+
+namespace {
+using ::testing::Return;
+
+const unsigned char* MockGetString(GLenum name) {
+  switch (name) {
+    case GL_VERSION:
+    case GL_SHADING_LANGUAGE_VERSION:
+      return (unsigned char*)"3.0";
+    default:
+      return (unsigned char*)"";
+  }
+}
+
+void MockGetIntegerv(GLenum name, int* value) {
+  *value = 0;
+}
+
+GLenum MockGetError() {
+  return GL_NO_ERROR;
+}
+
+void DoNothing() {}
+
+const impeller::ProcTableGLES::Resolver kMockResolver = [](const char* name) {
+  if (strcmp(name, "glGetString") == 0) {
+    return reinterpret_cast<void*>(&MockGetString);
+  } else if (strcmp(name, "glGetIntegerv") == 0) {
+    return reinterpret_cast<void*>(&MockGetIntegerv);
+  } else if (strcmp(name, "glGetError") == 0) {
+    return reinterpret_cast<void*>(&MockGetError);
+  } else {
+    return reinterpret_cast<void*>(&DoNothing);
+  }
+};
+
+class MockAngleSurfaceManager : public AngleSurfaceManager {
+ public:
+  MockAngleSurfaceManager() : AngleSurfaceManager(false) {}
+
+  MOCK_METHOD(bool, MakeCurrent, (), (override));
+
+ private:
+  FML_DISALLOW_COPY_AND_ASSIGN(MockAngleSurfaceManager);
+};
+
+class MockFlutterWindowsView : public FlutterWindowsView {
+ public:
+  MockFlutterWindowsView(std::unique_ptr<WindowBindingHandler> window)
+      : FlutterWindowsView(std::move(window)) {}
+  virtual ~MockFlutterWindowsView() = default;
+
+  MOCK_METHOD(bool, SwapBuffers, (), (override));
+
+ private:
+  FML_DISALLOW_COPY_AND_ASSIGN(MockFlutterWindowsView);
+};
+
+class CompositorOpenGLTest : public WindowsTest {
+ public:
+  CompositorOpenGLTest() = default;
+  virtual ~CompositorOpenGLTest() = default;
+
+ protected:
+  FlutterWindowsEngine* engine() { return engine_.get(); }
+  MockFlutterWindowsView* view() { return view_.get(); }
+  MockAngleSurfaceManager* surface_manager() { return surface_manager_; }
+
+  void UseHeadlessEngine() {
+    auto surface_manager = std::make_unique<MockAngleSurfaceManager>();
+    surface_manager_ = surface_manager.get();
+
+    FlutterWindowsEngineBuilder builder{GetContext()};
+
+    engine_ = builder.Build();
+    EngineModifier modifier(engine_.get());
+    modifier.SetSurfaceManager(surface_manager.release());
+  }
+
+  void UseEngineWithView() {
+    UseHeadlessEngine();
+
+    auto window = std::make_unique<MockWindowBindingHandler>();
+    EXPECT_CALL(*window.get(), SetView).Times(1);
+    EXPECT_CALL(*window.get(), GetRenderTarget).WillRepeatedly(Return(nullptr));
+
+    view_ = std::make_unique<MockFlutterWindowsView>(std::move(window));
```

Comment:
Question -- why use the local? This assignment is inevitable since we're using `EXPECT_CALL`, so the test will continue running even if either of those checks fails.

Reply:
I'm not sure I understand this feedback. These `EXPECT_CALL`s are necessary mocks for `FlutterWindowsView`'s constructor and need to be registered before we create the `FlutterWindowsView`. Please let me know if I misunderstood your question.

Reply:
Ah sorry - my bad. I missed that you're using a subtype here so you'd need narrowing casts if you made the assignment to `view_` directly without the local.